compiler/circle-mpqsolver/src/BisectionSolver.cpp

```diff
@@ -0,0 +1,220 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "BisectionSolver.h"
+#include "ErrorApproximator.h"
+#include "ModuleCloner.h"
+#include "Evaluator.h"
+#include "DepthParameterizer.h"
+
+#include <cmath>
+#include <iostream>
+
+using namespace mpqsolver;
+
+namespace
+{
+
+bool error_at_input_is_larger_than_at_output(const NodeDepthType &nodes_depth, float cut_depth)
+{
+  float error_at_input = 0;
+  float error_at_output = 0;
+  for (auto &iter : nodes_depth)
+  {
+    float cur_error = ErrorApproximator::approximate(iter.first);
+    if (iter.second < cut_depth)
+    {
+      error_at_input += cur_error;
+    }
+    else
+    {
+      error_at_output += cur_error;
+    }
+  }
+
+  if (error_at_input > error_at_output)
+  {
+    std::cerr << "Q16 will be set at input due to ";
+  }
+  else
+  {
+    std::cerr << "Q8 will be set at input due to ";
+  }
+  std::cerr << error_at_input << " error at input vs ";
+  std::cerr << error_at_output << " error at output." << std::endl;
+
+  return error_at_input > error_at_output;
+}
+
+class BisectionOptionsImpl final : public BisectionSolver::Options
+{
+public:
+  virtual void enable(Q16AtInput) final;
+  virtual bool query(Q16AtInput) final;
+
+private:
+  Q16AtInput _q16AtInput = Q16AtInput::Auto;
+};
+
+void BisectionOptionsImpl::enable(Q16AtInput q16AtInput) { _q16AtInput = q16AtInput; }
+bool BisectionOptionsImpl::query(Q16AtInput q16AtInput) { return _q16AtInput == q16AtInput; }
+
+} // namespace
+
+BisectionSolver::BisectionSolver(const std::string &input_data_path, float qerror_ratio)
+  : _input_data_path(input_data_path), _qerror_ratio(qerror_ratio), _qerror(0.f)
+{
+}
+
+BisectionSolver::Options *BisectionSolver::options(void)
+{
+  if (_options == nullptr)
+  {
+    _options = std::make_unique<BisectionOptionsImpl>();
+  }
+
+  return _options.get();
+}
+
+std::unique_ptr<luci::Module> BisectionSolver::run(const luci::Module *in_module)
+{
+  float min_depth = 0.f;
+  float max_depth = 0.f;
+  NodeDepthType nodes_depth;
+  if (DepthParameterizer::compute_depth(in_module, nodes_depth, min_depth, max_depth) !=
+      EXIT_SUCCESS)
+  {
+    std::cerr << "Invalid graph for bisectioning" << std::endl;
+    return nullptr;
+  }
+
+  DatasetEvaluator evaluator(in_module, _input_data_path);
+  // const auto &ref_output = compute_outputs(module.get(), _input_data_path);
+  LayerParams layer_params;
+  float int16_qerror = evaluator.evaluate("int16", layer_params);
+  std::cerr << "int16_quantization_error " << int16_qerror << std::endl;
```

Comment:
Q) will this message be seen without condition? is this really a error or log?
if it's a log, please refer other codes in `compiler` like _luci_ and use something like `LOGGER()`

Reply:
Ahhh...Ok.